File: src/platform/CloudPlatform.cpp
#include "platform/CloudPlatform.hpp"
#include <sstream>

namespace enigma {

ZoneConfig CloudPlatform::createDataCenter(int numRacks, int serversPerRack,
                                            const std::string& serverSpeed) {
    ZoneConfig zone("data_center", "Full");
    
    // Create racks
    for (int r = 0; r < numRacks; ++r) {
        ZoneConfig rack("rack_" + std::to_string(r), "Full");
        
        // Servers in each rack
        for (int s = 0; s < serversPerRack; ++s) {
            std::string id = "server_r" + std::to_string(r) + "_s" + std::to_string(s);
            rack.hosts.emplace_back(id, serverSpeed, 32);
        }
        
        // Intra-rack link (very high speed)
        rack.links.emplace_back("rack_switch", "40GBps", "10us");
        
        zone.subzones.push_back(rack);
    }
    
    // Inter-rack links (high speed, higher latency)
    zone.links.emplace_back("spine_switch", "10GBps", "50us");
    
    return zone;
}

ZoneConfig CloudPlatform::createCluster(int numServers,
                                        const std::string& serverSpeed,
                                        const std::string& interconnect) {
    ZoneConfig zone("cloud_cluster", "Full");
    
    // Create homogeneous servers
    for (int i = 0; i < numServers; ++i) {
        std::string id = "server_" + std::to_string(i);
        zone.hosts.emplace_back(id, serverSpeed, 32);
    }
    
    // High-speed interconnect
    zone.links.emplace_back("cluster_interconnect", interconnect, "100us");
    
    return zone;
}

ZoneConfig CloudPlatform::createMultiCloud(int numClouds, int serversPerCloud,
                                           const std::string& serverSpeed) {
    ZoneConfig zone("multi_cloud", "Full");
    
    // Create multiple clouds
    for (int c = 0; c < numClouds; ++c) {
        ZoneConfig cloud("cloud_" + std::to_string(c), "Full");
        
        // Servers in each cloud
        for (int s = 0; s < serversPerCloud; ++s) {
            std::string id = "cloud" + std::to_string(c) + "_server_" + std::to_string(s);
            cloud.hosts.emplace_back(id, serverSpeed, 32);
        }
        
        // Intra-cloud link
        cloud.links.emplace_back("intra_cloud_link", "10GBps", "100us");
        
        zone.subzones.push_back(cloud);
    }
    
    // Inter-cloud links (WAN)
    zone.links.emplace_back("inter_cloud_link", "1GBps", "50ms");
    
    return zone;
}

ZoneConfig CloudPlatform::createHeterogeneousCluster(int numCPUNodes, int numGPUNodes,
                                                      const std::string& cpuSpeed,
                                                      const std::string& gpuSpeed) {
    ZoneConfig zone("heterogeneous_cluster", "Full");
    
    // CPU nodes
    for (int i = 0; i < numCPUNodes; ++i) {
        std::string id = "cpu_node_" + std::to_string(i);
        zone.hosts.emplace_back(id, cpuSpeed, 64);
    }
    
    // GPU nodes (higher compute capacity)
    for (int i = 0; i < numGPUNodes; ++i) {
        std::string id = "gpu_node_" + std::to_string(i);
        zone.hosts.emplace_back(id, gpuSpeed, 128);
    }
    
    // High-performance interconnect (InfiniBand)
    zone.links.emplace_back("infiniband", "100GBps", "1us");
    
    return zone;
}

HostConfig CloudPlatform::createCloudServer(const std::string& id,
                                            const std::string& serverType) {
    if (serverType == "small") {
        return HostConfig(id, "50Gf", 8);
    } else if (serverType == "medium") {
        return HostConfig(id, "100Gf", 16);
    } else if (serverType == "large") {
        return HostConfig(id, "200Gf", 32);
    } else if (serverType == "xlarge") {
        return HostConfig(id, "400Gf", 64);
    } else if (serverType == "gpu") {
        return HostConfig(id, "1000Gf", 128);
    } else if (serverType == "hpc") {
        return HostConfig(id, "500Gf", 128);
    } else {
        // Standard
        return HostConfig(id, "100Gf", 32);
    }
}

LinkConfig CloudPlatform::createCloudLink(const std::string& id,
                                          const std::string& linkType) {
    if (linkType == "1G") {
        return LinkConfig(id, "1GBps", "500us");
    } else if (linkType == "10G") {
        return LinkConfig(id, "10GBps", "100us");
    } else if (linkType == "40G") {
        return LinkConfig(id, "40GBps", "50us");
    } else if (linkType == "100G") {
        return LinkConfig(id, "100GBps", "10us");
    } else if (linkType == "infiniband") {
        return LinkConfig(id, "200GBps", "1us");
    } else {
        // Default 10G
        return LinkConfig(id, "10GBps", "100us");
    }
}

} // namespace enigma

File: src/python/enigma/platform/cloud_platform.py
"""
CloudPlatform: specialized factories for cloud computing platform topologies.
Mirrors C++ CloudPlatform class from src/platform/CloudPlatform.cpp.
"""

from __future__ import annotations

from .configs import HostConfig, LinkConfig, ZoneConfig


class CloudPlatform:
    """Factory for cloud computing platform topologies."""

    # ------------------------------------------------------------------ #
    # Topology factories                                                   #
    # ------------------------------------------------------------------ #

    @staticmethod
    def create_data_center(
        num_racks: int,
        servers_per_rack: int,
        server_speed: str = "100Gf",
    ) -> ZoneConfig:
        """Fat-tree-style data center: racks of servers connected via spine switch."""
        zone = ZoneConfig("data_center", "Full")
        for r in range(num_racks):
            rack = ZoneConfig(f"rack_{r}", "Full")
            for s in range(servers_per_rack):
                rack.hosts.append(
                    HostConfig(f"server_r{r}_s{s}", server_speed, 32)
                )
            rack.links.append(LinkConfig("rack_switch", "40GBps", "10us"))
            zone.subzones.append(rack)
        zone.links.append(LinkConfig("spine_switch", "10GBps", "50us"))
        return zone

    @staticmethod
    def create_cluster(
        num_servers: int,
        server_speed: str = "100Gf",
        interconnect: str = "10GBps",
    ) -> ZoneConfig:
        """Simple homogeneous cloud cluster."""
        zone = ZoneConfig("cloud_cluster", "Full")
        for i in range(num_servers):
            zone.hosts.append(HostConfig(f"server_{i}", server_speed, 32))
        zone.links.append(LinkConfig("cluster_interconnect", interconnect, "100us"))
        return zone

    @staticmethod
    def create_multi_cloud(
        num_clouds: int,
        servers_per_cloud: int,
        server_speed: str = "100Gf",
    ) -> ZoneConfig:
        """Multiple independent cloud regions connected via WAN links."""
        zone = ZoneConfig("multi_cloud", "Full")
        for c in range(num_clouds):
            cloud = ZoneConfig(f"cloud_{c}", "Full")
            for s in range(servers_per_cloud):
                cloud.hosts.append(
                    HostConfig(f"cloud{c}_server_{s}", server_speed, 32)
                )
            cloud.links.append(LinkConfig("intra_cloud_link", "10GBps", "100us"))
            zone.subzones.append(cloud)
        zone.links.append(LinkConfig("inter_cloud_link", "1GBps", "50ms"))
        return zone

    @staticmethod
    def create_heterogeneous_cluster(
        num_cpu_nodes: int,
        num_gpu_nodes: int,
        cpu_speed: str = "100Gf",
        gpu_speed: str = "500Gf",
    ) -> ZoneConfig:
        """Cluster with both CPU and GPU nodes, connected via InfiniBand."""
        zone = ZoneConfig("heterogeneous_cluster", "Full")
        for i in range(num_cpu_nodes):
            zone.hosts.append(HostConfig(f"cpu_node_{i}", cpu_speed, 64))
        for i in range(num_gpu_nodes):
            zone.hosts.append(HostConfig(f"gpu_node_{i}", gpu_speed, 128))
        zone.links.append(LinkConfig("infiniband", "100GBps", "1us"))
        return zone

    # ------------------------------------------------------------------ #
    # Individual component factories                                       #
    # ------------------------------------------------------------------ #

    @staticmethod
    def create_cloud_server(server_id: str, server_type: str = "standard") -> HostConfig:
        presets = {
            "small":    HostConfig(server_id, "10Gf",  8),
            "standard": HostConfig(server_id, "100Gf", 32),
            "large":    HostConfig(server_id, "500Gf", 64),
            "gpu":      HostConfig(server_id, "1000Gf", 128),
        }
        return presets.get(server_type, HostConfig(server_id, "100Gf", 32))

    @staticmethod
    def create_cloud_link(link_id: str, link_type: str = "10G") -> LinkConfig:
        presets = {
            "1G":   LinkConfig(link_id, "1GBps",   "100us"),
            "10G":  LinkConfig(link_id, "10GBps",  "50us"),
            "40G":  LinkConfig(link_id, "40GBps",  "10us"),
            "100G": LinkConfig(link_id, "100GBps", "1us"),
            "wan":  LinkConfig(link_id, "1GBps",   "50ms"),
        }
        return presets.get(link_type, LinkConfig(link_id, "10GBps", "50us"))
